db/sqldb/portfolio.go

```diff
@@ -3,45 +3,50 @@ package sqldb
 import (
 	"errors"
 	"fmt"
+	"time"
 
 	"github.com/RMI/pacta/db"
 	"github.com/RMI/pacta/pacta"
 	"github.com/jackc/pgx/v5"
 	"github.com/jackc/pgx/v5/pgtype"
 )
 
-const portfolioSelectColumns = `
-	portfolio.id,
-	portfolio.owner_id,
-	portfolio.name,
-	portfolio.description,
-	portfolio.created_at,
-	portfolio.holdings_date,
-	portfolio.blob_id,
-	portfolio.admin_debug_enabled,
-	portfolio.number_of_rows 
+const portfolioQueryStanza = `
+	SELECT
+		portfolio.id,
+		portfolio.owner_id,
+		portfolio.name,
+		portfolio.description,
+		portfolio.created_at,
+		portfolio.holdings_date,
+		portfolio.blob_id,
+		portfolio.admin_debug_enabled,
+		portfolio.number_of_rows,
+		portfolio_group_membership.portfolio_group_id,
+		portfolio_group_membership.created_at
+	FROM portfolio
+	LEFT JOIN portfolio_group_membership 
+	ON portfolio_group_membership.portfolio_id = portfolio.id
 `
 
 func (d *DB) Portfolio(tx db.Tx, id pacta.PortfolioID) (*pacta.Portfolio, error) {
-	rows, err := d.query(tx, `
-		SELECT `+portfolioSelectColumns+`
-		FROM portfolio 
-		WHERE id = $1;`, id)
+	rows, err := d.query(tx, portfolioQueryStanza+` WHERE id = $1;`, id)
 	if err != nil {
 		return nil, fmt.Errorf("querying portfolio: %w", err)
 	}
 	pvs, err := rowsToPortfolios(rows)
 	if err != nil {
 		return nil, fmt.Errorf("translating rows to portfolios: %w", err)
 	}
-	return exactlyOne("portfolio", id, pvs)
+	return exactlyOne("portfolio", id, valuesFromMap(pvs))
```

Comment:
I had initially complained that it seemed silly to query by primary key with `query` instead of `queryRow`, only to then verify we got a single entry. Now I'm doubly complaining because we're loading what we know is zero or one result(s) into a map, to load it into an array, to verify we got a single entry. Not that this is anywhere near a bottleneck, but it would be simpler and more efficient just to use queryRow and check for ErrNoRows or whatever it is.

Reply:
I agree with the diagnosis of layercaking here, but the benefit IMO outweighs the extra layer. One code pathway for decoding rows allows us to build code that is (a) more resilient to changes (i.e. we couldn't read a field in one case but not in another), (b) less work to maintain (i.e. a change to the data model needs to be made in fewer places)